### PR TITLE
[Profiler][Minor] Set end time on python events when profiling stops.

### PR DESCRIPTION
**Pull request by robieta**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83622
* __->__ #83621
* #83033
* #82993
* #82992

We don't have an end event for calls that are ongoing when profiling stops. (e.g. main) This cropped up when I was adding checks for negative durations.

I also refactored `populate` to use a pop method. This not only allows me to implement this fix, but should also provide a convenient entry point for https://github.com/pytorch/pytorch/pull/82154

Differential Revision: [D38426342](https://our.internmc.facebook.com/intern/diff/D38426342/)

### CODE REVIEW

**Review by slgong-fb — APPROVED** [review]
Looks good to me! Need to take care of CI tests that are currently failing.

**Comment by robieta** [comment]
> Looks good to me! Need to take care of CI tests that are currently failing.

I think those are from an issue with an earlier PR, but we'll see. I also realized I should only be adding `SOFT_ASSERT`s, not any more `TORCH_INTERNAL_ASSERT`s.

File: torch/csrc/profiler/collection.cpp [hunk 1]
```diff
@@ -447,13 +448,21 @@ uint64_t Result::correlationID() const {
...
-  return visit(c10::overloaded(
+  auto end_time_ns = visit(c10::overloaded(
       ATTRIBUTE(TorchOp, torchOpEndNS(e, finished_, parent_)),
       ATTRIBUTE(Backend, e.end_time_us_ * 1000),
       ATTRIBUTE(Allocation, start_time_ns_),
       ATTRIBUTE(OutOfMemory, start_time_ns_),
       ATTRIBUTE(Kineto, start_time_ns_ + e.duration_us_ * 1000),
       [&](const auto& e) -> int64_t { return e.end_time_ns_; }));
+
+  // In rare cases we're willing to tolerate ops which are missing an end time
```
Comment (zejun-chen) [inline torch/csrc/profiler/collection.cpp:459]: Hi @robieta 
May i know what the rare cases are?

**Pull request closed without merging**